go/proxyd/methods.go

```diff
@@ -193,3 +251,98 @@ func validBlockInput(input string) bool {
 	_, err := decodeBlockInput(input)
 	return err == nil
 }
+
+func makeRPCRes(req *RPCReq, result interface{}) *RPCRes {
+	return &RPCRes{
+		JSONRPC: JSONRPCVersion,
+		ID:      req.ID,
+		Result:  result,
+	}
+}
+
+func copyResError(err *RPCErr) *RPCErr {
+	if err == nil {
+		return nil
+	}
+	return &RPCErr{
+		Code:          err.Code,
+		Message:       err.Message,
+		HTTPErrorCode: err.HTTPErrorCode,
+	}
+}
+
+func copyRes(res *RPCRes) *RPCRes {
+	return &RPCRes{
+		JSONRPC: res.JSONRPC,
+		Result:  res.Result,
+		Error:   copyResError(res.Error),
+		ID:      res.ID,
+	}
+}
+
+type CachedRPC struct {
+	BlockNum   uint64  `json:"blockNum"`
+	Res        *RPCRes `json:"res"`
+	Expiration int64   `json:"expiration"` // in millis since epoch
+}
+
+func (c *CachedRPC) Encode() []byte {
+	return mustMarshalJSON(c)
+}
+
+func (c *CachedRPC) Decode(b []byte) error {
+	return json.Unmarshal(b, c)
+}
+
+func (c *CachedRPC) ExpirationTime() time.Time {
+	return time.Unix(0, c.Expiration*int64(time.Millisecond))
+}
+
+func getBlockDependentCachedRPCResponse(ctx context.Context, cache Cache, getLatestBlockNumFn GetLatestBlockNumFn, key string, req *RPCReq) (*RPCRes, error) {
+	encodedVal, err := cache.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	if encodedVal == "" {
+		return nil, nil
+	}
+	val, err := snappy.Decode(nil, []byte(encodedVal))
+	if err != nil {
+		return nil, err
+	}
+
+	item := new(CachedRPC)
+	if err := json.Unmarshal(val, item); err != nil {
+		return nil, err
+	}
+	curBlockNum, err := getLatestBlockNumFn(ctx)
+	if err != nil {
+		return nil, err
+	}
+	expired := time.Now().After(item.ExpirationTime())
+	if (curBlockNum > item.BlockNum && expired) ||
+		(curBlockNum < item.BlockNum) /* desync: reorgs, backend failover */ {
+		// TODO: be careful removing keys once there are multiple proxyd instances
+		err := cache.Remove(ctx, key)
+		return nil, err
+	}
+
+	res := item.Res
+	res.ID = req.ID
+	return res, nil
+}
+
+func putBlockDependentCachedRPCResponse(ctx context.Context, cache Cache, key string, res *RPCRes, blockNumberSync uint64) error {
+	if key == "" {
+		return nil
+	}
+	item := CachedRPC{
+		BlockNum:   blockNumberSync,
+		Res:        res,
+		Expiration: time.Now().Add(cacheTTL).UnixNano() / int64(time.Millisecond),
```

Comment:
The redis key ttl should outlive the Expiration (probably by a factor of 2). Best to allow redis expire keys only after we're totally done with them.